pilot/cmd/pilot-agent/main_test.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"github.com/onsi/gomega"
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/pilot/pkg/serviceregistry"
+	"os"
+	"testing"
+)
+
+func TestNoPilotSanIfAuthenticationNone(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	role.DNSDomain = ""
+	role.TrustDomain = ""
+	controlPlaneAuthPolicy = meshconfig.AuthenticationPolicy_NONE.String()
+
+	pilotSAN := getPilotSAN(role.DNSDomain, "anything")
+
+	g.Expect(pilotSAN).To(gomega.BeNil())
+}
+
+func TestPilotSanIfAuthenticationMutualDomainEmptyKubernetes(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	role.DNSDomain = ""
+	role.TrustDomain = ""
+	registry = serviceregistry.KubernetesRegistry
+	controlPlaneAuthPolicy = meshconfig.AuthenticationPolicy_MUTUAL_TLS.String()
+
+	pilotSAN := getPilotSAN(role.DNSDomain, "anything")
+
+	g.Expect(pilotSAN).To(gomega.Equal([]string{"spiffe://cluster.local/ns/anything/sa/istio-pilot-service-account"}))
+}
+
+func TestPilotSanIfAuthenticationMutualDomainNotEmptyKubernetes(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	role.DNSDomain = "my.domain"
+	role.TrustDomain = ""
+	registry = serviceregistry.KubernetesRegistry
+	controlPlaneAuthPolicy = meshconfig.AuthenticationPolicy_MUTUAL_TLS.String()
+
+	pilotSAN := getPilotSAN(role.DNSDomain, "anything")
+
+	g.Expect(pilotSAN).To(gomega.Equal([]string{"spiffe://my.domain/ns/anything/sa/istio-pilot-service-account"}))
+}
+
+//TODO Is this really correct?
+func TestPilotSanIfAuthenticationMutualDomainEmptyConsul(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	role.DNSDomain = ""
+	role.TrustDomain = ""
+	registry = serviceregistry.ConsulRegistry
+	controlPlaneAuthPolicy = meshconfig.AuthenticationPolicy_MUTUAL_TLS.String()
+
+	pilotSAN := getPilotSAN(role.DNSDomain, "anything")
+
+	g.Expect(pilotSAN).To(gomega.Equal([]string{"spiffe:///ns/anything/sa/istio-pilot-service-account"}))
```

Comment:
FYI, this is not a valid SPIFFE ID since SPIFFE requires a non-empty trust domain. See https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md#2-spiffe-identity.

If the intention here is to test against an invalid SPIFFE URI, I suggest making that intention explicit in a comment.

Reply:
This test was used to ensure that the former behaviour is not changed.
We will add a comment to the test